app/models/grda_warehouse/ch_enrollment.rb

```diff
@@ -74,5 +74,323 @@ def self.update_existing!
         )
       end
     end
+
+    # Accept an optional date which will be used for extending the homeless
+    # range if the project is a homeless project
+    def self.chronically_homeless_at_start?(enrollment, date: enrollment.EntryDate)
+      chronically_homeless_at_start(enrollment, date: date) == :yes
+    end
+
+    # Was the client chronically homeless at the start of this enrollment?
+    # Optionally accepts a date to use for "CH at a point-in-time" calculation.
+    #
+    # @return [Symbol] :yes, :no, :dk_or_r, or :missing
+    def self.chronically_homeless_at_start(enrollment, date: enrollment.EntryDate)
+      chronically_homeless_at_start_steps(enrollment, date: date).last[:result]
+    end
+
+    # Was the client chronically homeless at the start of this enrollment?
+    # Optionally accepts a date to use for "CH at a point-in-time" calculation.
+    #
+    # @return [Array] all steps evaluated with result and display values.
+    #
+    # Each step in the array has this shape:
+    # {
+    #    result: Symbol, true, false, or nil
+    #    display_value: value to display in chronic-at-entry explanation table
+    #    line: [Number] line number in HUD calculation
+    # }
+    #
+    # Result "nil" means continue processing
+    # Result "false" means continue processing and skip branch
+    # Result "true" means continue processing and enter branch
+    # The last item in the array will have one of (:yes, :no, :dk_or_r, or :missing) as the result
+    def self.chronically_homeless_at_start_steps(enrollment, date: enrollment.EntryDate)
```

Comment:
This main function is updated to return an array of executed steps, rather than just the final result. We have pretty thorough tests for the function, it looks like, so I am hopeful I didn't mess anything up. I spot checked a bunch of enrollments (screenshots attached of some) and it looks correct to me.